frontend/lib/src/components/widgets/ButtonGroup/ButtonGroup.tsx

```diff
@@ -0,0 +1,334 @@
+ */
+function showAsSelected(
+  selectionVisualization: ButtonGroupProto.SelectionVisualization,
+  clickMode: ButtonGroupProto.ClickMode,
+  selected: number[],
+  index: number
+): boolean {
+  if (selected.indexOf(index) > -1) {
+    return true
+  }
+
+  if (
+    clickMode !== ButtonGroupProto.ClickMode.SINGLE_SELECT ||
+    selectionVisualization !==
+      ButtonGroupProto.SelectionVisualization.ALL_UP_TO_SELECTED
+  ) {
+    return false
+  }
+
+  return selected.length > 0 && index < selected[0]
+}
+
+function getContent(
+  isVisuallySelected: boolean,
+  fallbackContent: string,
+  selectionContent?: string | undefined | null
+): string {
+  if (isVisuallySelected && selectionContent) {
+    return selectionContent
+  }
+
+  return fallbackContent
+}
+
+function getSelectedStyle(
+  isVisuallySelected: boolean,
+  isSelectionHighlightDisabled: boolean,
+  theme: EmotionTheme
+): Record<string, string> | undefined {
+  if (!isVisuallySelected || isSelectionHighlightDisabled) {
+    return undefined
+  }
+
+  return {
+    backgroundColor: theme.colors.lightGray,
+  }
+}
+
+function createOptionChild(
+  option: ButtonGroupProto.IOption,
+  index: number,
+  selectionVisualization: ButtonGroupProto.SelectionVisualization,
+  clickMode: ButtonGroupProto.ClickMode,
+  selected: number[],
+  theme: EmotionTheme
+): React.FunctionComponent {
+  const isVisuallySelected = showAsSelected(
+    selectionVisualization,
+    clickMode,
+    selected,
+    index
+  )
+  const content = getContent(
+    isVisuallySelected,
+    option.content ?? "",
+    option.selectedContent
+  )
+  const style = getSelectedStyle(
+    isVisuallySelected,
+    option.disableSelectionHighlight || false,
+    theme
+  )
+
+  // we have to use forwardRef here becaused BasewebButtonGroup passes it down to its children
+  return forwardRef(function BaseButtonGroup(
+    props: any,
+    _: Ref<BasewebButtonGroup>
+  ): ReactElement {
+    return (
+      <BaseButton
+        {...props}
+        size={BaseButtonSize.SMALL}
+        kind={BaseButtonKind.BUTTON_GROUP}
+        style={style}
+      >
+        {getContentElement(content)}
+      </BaseButton>
+    )
+  })
+}
+
+function getInitialValue(
+  widgetMgr: WidgetStateManager,
+  element: ButtonGroupProto
+): number[] {
+  const storedValue = widgetMgr.getIntArrayValue(element)
+  return storedValue ?? element.default
+}
+
+function ButtonGroup(props: Readonly<Props>): ReactElement {
+  const { disabled, element, fragmentId, widgetMgr } = props
+  const {
+    clickMode,
+    default: defaultValues,
+    options,
+    setValue,
+    value,
+    selectionVisualization,
+  } = element
+
+  const theme: EmotionTheme = useTheme()
+
+  const [selected, setSelected] = useState<number[]>(
+    getInitialValue(widgetMgr, element) || []
+  )
+  // initial render is due to mounting, hence setting it to false
+  const [isSetFromUi, setFromUi] = useState<boolean>(false)
```

Comment:
Instead of this flag, I would just check if the new value is actually different from the one currently in the state and only sync in that case (within `syncValue`). With this logic, we also prevent the rerun if you click on an already selected option. I don't think we want to trigger a rerun in that case: 

https://github.com/streamlit/streamlit/assets/2852129/0ccf2ca9-c8b6-42ba-9b27-e94f1938076d

Reply:
So you mean we should not sync when the component mounts? If yes, then this would be a change to how other select elements handle this today, see [here](https://github.com/streamlit/streamlit/blob/develop/frontend/lib/src/components/widgets/Multiselect/Multiselect.tsx#L103-L109).

Reply:
But the linked example also does not sync to the backend when the component mounts since `fromUi` is false? It just commits it to the frontend widget state.

https://github.com/streamlit/streamlit/blob/a81b9c05f07a6962d9e092fa2d200452f11c4fa8/frontend/lib/src/components/widgets/Multiselect/Multiselect.tsx#L107

Reply:
ah great point, I somehow thought that this is just a signal flag. Do we need to commit to the frontend widget state?

Reply:
yeah, not 100% sure if this is really required but committing just to the frontend state might not hurt.

Reply:
Updated it based on your suggestion 👍